File: app/src/main/java/ru/bloknot/news/adapters/CustomAdapter.java
package ru.bloknot.news.adapters;

import android.content.Context;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.squareup.picasso.Picasso;

import java.util.List;

import ru.bloknot.news.R;
import ru.bloknot.news.activity.FullActivity;
import ru.bloknot.news.models.CardNews;

public class CustomAdapter extends RecyclerView.Adapter<CustomAdapter.ViewHolder> {
    private final List<CardNews> dataList;
    private final Context context;

    public CustomAdapter(Context context, List<CardNews> dataList) {
        this.context = context;
        this.dataList = dataList;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.my_test, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        CardNews data = dataList.get(position);
        holder.title.setText(data.getTitle());
        holder.category.setText(data.getCategory());
        holder.time.setText(data.getTime());
        holder.description.setText(data.getDescription());
        Picasso.get()
                .load(data.getImageUrl())
                .fit()
                .into(holder.imageView);

        holder.itemView.setOnClickListener(view -> {
            int pos = holder.getAdapterPosition();
            System.out.println("клик: " + pos);
            Intent intent = new Intent(view.getContext(), FullActivity.class);
            intent.putExtra("count", pos);
            view.getContext().startActivity(intent);
        });
    }

    @Override
    public int getItemCount() {
        return dataList.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        TextView title, category, time, description;
        ImageView imageView;
        public ViewHolder(View itemView) {
            super(itemView);
            title = itemView.findViewById(R.id.titleText);
            category = itemView.findViewById(R.id.catText);
            time = itemView.findViewById(R.id.timeText);
            description = itemView.findViewById(R.id.descriptionText);
            imageView = itemView.findViewById(R.id.imageView);
        }
    }
}

File: app/src/main/java/ru/bloknot/news/adapters/CustomAdapter.kt
package ru.bloknot.news.adapters

import android.content.Intent
import android.util.Log
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import coil.load
import coil.request.CachePolicy
import coil.size.Scale
import ru.bloknot.news.R
import ru.bloknot.news.activity.FullActivity
import ru.bloknot.news.databinding.MyTestBinding
import ru.bloknot.news.models.CardNews

class CustomAdapter(
    private val newsList: List<CardNews>,
) : RecyclerView.Adapter<CustomAdapter.NewsViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): NewsViewHolder {
        val binding = MyTestBinding.inflate(
            LayoutInflater.from(parent.context),
            parent,
            false
        )
        return NewsViewHolder(binding)
    }

    override fun onBindViewHolder(holder: NewsViewHolder, position: Int) {
        holder.bind(newsList[position])
    }

    override fun getItemCount(): Int = newsList.size

    // ViewHolder с ViewBinding — никаких findViewById!
    class NewsViewHolder(
        private val binding: MyTestBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        fun bind(item: CardNews) {
            binding.titleText.text = item.getTitle()
            binding.catText.text = item.getCategory()
            binding.timeText.text = item.getTime()
            binding.descriptionText.text = item.getDescription()
            val urlFull = item.fullUrlLink

            binding.root.setOnClickListener {
                Log.d("CLICK_DEBUG", "Клик по новости: ${item.getTitle()}")
                Log.d("CLICK_DEBUG", "URL перед отправкой: ${item.getFullUrlLink() ?: "NULL или пусто!"}")

                val context = binding.root.context
                val intent = Intent(context, FullActivity::class.java)
                intent.putExtra("url", urlFull)
                intent.putExtra("title", binding.titleText.text)
                context.startActivity(intent)
            }

            // Coil — лучшая библиотека для загрузки картинок в 2025 году
            binding.imageView.load(item.imageUrl) {
                crossfade(false)                          // ← отключаем анимацию появления
                placeholder(R.drawable.lazy)
                error(R.drawable.error)
                scale(Scale.FILL)                         // ← растягивает на весь размер (аналог centerCrop, но без обрезки по центру)
                allowHardware(false)                      // отключает hardware acceleration для стабильности
                memoryCachePolicy(CachePolicy.ENABLED)    // кэшируем в памяти
                diskCachePolicy(CachePolicy.ENABLED)      // кэшируем на диск
            }
        }
    }
}